### ISSUE TITLE
residmaps broken in 0.19.0+7.gd54d.dirty 

### ISSUE THREAD

**Issue by me-manu** [fermiPy/fermipy#345]
I just pulled the latest fermipy version (0.19.0+7.gd54d.dirty) and now my residual map computation breaks with the error that my `GTanalysis` object does not have an attribute `binsz`:
```
    192         gta.residmap(modelname, model=model3,
    193                      loge_bounds=loge_bounds,
--> 194                      make_plots=True)
    195         gta.tsmap(modelname, model=model_pl20,
    196                                     loge_bounds=loge_bounds, make_plots=True)

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in residmap(self, prefix, **kwargs)
    256         config['model'].setdefault('Prefactor', 1E-13)
    257 
--> 258         o = self._make_residual_map(prefix, **config)
    259 
    260         if config['make_plots']:

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in _make_residual_map(self, prefix, **kwargs)
    308             return self._make_residual_map_hpx(prefix, **kwargs)
    309         elif self.projtype == "WCS":
--> 310             return self._make_residual_map_wcs(prefix, **kwargs)
    311         else:
    312             raise Exception(

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in _make_residual_map_wcs(self, prefix, **kwargs)
    350         if src_dict['SpatialModel'] == 'Gaussian':
    351             kernel = utils.make_gaussian_kernel(src_dict['SpatialWidth'],
--> 352                                                 cdelt=self.binsz,
    353                                                 npix=101)
    354             kernel /= np.sum(kernel)

AttributeError: 'GTAnalysis' object has no attribute 'binsz'
```
I tried a simple fix that add to add the property `binsz` to the GTanalysis object like so:
```
@property
def binsz(self):
    return self._binsz
```
around line 505 in gtanalysis.py. When I do that, however, the `get_source_kernel` function fails:
```
/u/gl/mmeyer/projects/python/fermiAnalysis/fermiAnalysis/ext_funcs.pyc in fit_region(gta, modelname, src_name, loge_bounds, skip_opt, shape_ts_threshold, force_ps, create_maps, create_sed, free_radius_sed, crate_ts_maps, distance_free_norm, distance_free_shape, **kwargs)
    192         gta.residmap(modelname, model=model3,
    193                      loge_bounds=loge_bounds,
--> 194                      make_plots=True)
    195         gta.tsmap(modelname, model=model_pl20,
    196                                     loge_bounds=loge_bounds, make_plots=True)

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in residmap(self, prefix, **kwargs)
    256         config['model'].setdefault('Prefactor', 1E-13)
    257 
--> 258         o = self._make_residual_map(prefix, **config)
    259 
    260         if config['make_plots']:

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in _make_residual_map(self, prefix, **kwargs)
    308             return self._make_residual_map_hpx(prefix, **kwargs)
    309         elif self.projtype == "WCS":
--> 310             return self._make_residual_map_wcs(prefix, **kwargs)
    311         else:
    312             raise Exception(

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in _make_residual_map_wcs(self, prefix, **kwargs)
    365         emst = np.zeros(self.npix[::-1])
    366 
--> 367         sm = get_source_kernel(self, 'residmap_testsource', kernel)
    368         ts = np.zeros(self.npix[::-1])
    369         sigma = np.zeros(self.npix[::-1])

/nfs/farm/g/glast/u/mmeyer/fermipy/fermipy/residmap.pyc in get_source_kernel(gta, name, kernel)
    197         z = c.model_counts_map(name).data.astype('float')
    198         if kernel is not None:
--> 199             shape = (z.shape[0][::-1],) + kernel.shape
    200             z = np.apply_over_axes(np.sum, z, axes=[1, 2]) * np.ones(
    201                 shape) * kernel[np.newaxis, :, :]

TypeError: 'int' object has no attribute '__getitem__'
```
since `z.shape[0][::-1]` does not work (and I don't understand what's supposed to be happening here). `z` is simply the data in the counts cube, so `z.shape` in my case is simply `(240, 240, 14)` and therefore `z.shape[0]` as an `int`.

**Comment by me-manu** [comment]
I just checked that before the rectangular roi pull request in line 199 was simply
`shape = (z.shape[0],) + kernel.shape`
and line 352 read
`cdelt=self.components[0].binsz,`
@tibaldo did you make these changes? It seems that now the residmap does not work with standard ROIs.

**Comment by tibaldo** [comment]
Hi @me-manu, thank you for reporting this. I made these changes, and I clearly made a mistake as you point out. It definitely should be `cdelt=self._binsz`.
I agree with you that it would be more logical to define the property `binsz`, which is probably why I made this mistake. Such property was not defined in the original version of gtanalysis.py, and I didn't add it myself when doing the modifications to the code.
If you look at #313 you will see that I had performed some tests, including residual map calculations, which worked fine, so there may be something subtle.
Furthermore, I find it really weird that this problem was not detected by the test suite.

**Comment by me-manu** [comment]
Thanks @tibaldo! But what about the error I get on line 199?

**Comment by tibaldo** [comment]
Hi @me-manu , I'm afraid I can't recall what led me to introduce that change. The fact this error went through, though, means that this part of the code was never run in my test, and, more importantly, in the test routine I guess.
If I understand correctly the code is supposed to build a cube with the same number of energy slices as the count map, and the spatial shape of the kernel. Is z.shape[0] the number of the energy planes in the cube indeed?
If that is the case this line should not have been modified at all to support the rectangular ROIs.

**Comment by me-manu** [comment]
Hi @tibaldo, 
I just checked and, indeed, `z.shape[0]` is the number of energy bins. I just pushed a new branch called `fix-345-residmaps` where I've included a `binsz` property and changed line 199 back to the old version. Could you check it out and see if it runs with your rectangular ROI?

**Comment by tibaldo** [comment]
I verified that with Manuel's changes basic analysis and residual map computation still works correctly in rectangular mode.
We also figured out why the problem was not detected by the unit tests. It turns out that test_gtanalysis.py was only testing residual map computation for the default point source model, while the error only occurred for Gaussian source models. @me-manu is adding a test for the latter case.
